lib/codegen/CreateTerm.cpp

```diff
@@ -99,100 +101,334 @@ llvm::Type *getValueType(SortCategory sort, llvm::Module *Module) {
-        llvm::StoreInst *StoreChild = new llvm::StoreInst(ChildValue, ChildPtr, block);
+        llvm::Value *ChildValue = (*this)(child);
+        llvm::Value *ChildPtr = llvm::GetElementPtrInst::CreateInBounds(BlockType, Block, {llvm::ConstantInt::get(llvm::Type::getInt64Ty(Ctx), 0), llvm::ConstantInt::get(llvm::Type::getInt32Ty(Ctx), idx++)}, "", CurrentBlock);
+        llvm::StoreInst *StoreChild = new llvm::StoreInst(ChildValue, ChildPtr, CurrentBlock);
       }
-      llvm::BitCastInst *Cast = new llvm::BitCastInst(Block, llvm::PointerType::get(Module->getTypeByName(BLOCK_STRUCT), 0), "", block);
+      llvm::BitCastInst *Cast = new llvm::BitCastInst(Block, llvm::PointerType::getUnqual(Module->getTypeByName(BLOCK_STRUCT)), "", CurrentBlock);
       return Cast;
     }
   } else {
     assert(false && "not supported yet: meta level");
   }
 }
 
+static int nextRuleId = 0;
+
+std::string makeApplyRuleFunction(KOREAxiomDeclaration *axiom, KOREDefinition *definition, llvm::Module *Module) {
+    KOREPattern *pattern = axiom->getRightHandSide();
+    llvm::StringMap<KOREObjectVariablePattern *> vars;
+    pattern->markVariables(vars);
+    llvm::StringMap<llvm::Type *> params;
+    std::vector<llvm::Type *> paramTypes;
+    std::vector<std::string> paramNames;
+    for (auto iter = vars.begin(); iter != vars.end(); ++iter) {
+      auto &entry = *iter;
+      auto sort = dynamic_cast<KOREObjectCompositeSort *>(entry.second->getSort());
+      if (!sort) {
+        // TODO: sort variables
+        return "";
+      }
+      llvm::Type *paramType = getValueType(sort->getCategory(definition), Module);
+      params.insert({entry.first(), paramType});
+      paramTypes.push_back(paramType);
+      paramNames.push_back(entry.first());
+    }
+    llvm::FunctionType *funcType = llvm::FunctionType::get(termType(pattern, params, definition, Module), paramTypes, false);
+    std::string name = "apply_rule_" + std::to_string(nextRuleId++);
+    llvm::Constant *func = Module->getOrInsertFunction(name, funcType);
+    llvm::Function *applyRule = llvm::cast<llvm::Function>(func);
+    llvm::StringMap<llvm::Value *> subst;
+    int i = 0;
+    for (auto val = applyRule->arg_begin(); val != applyRule->arg_end(); ++val, ++i) {
+      val->setName(paramNames[i]);
+      subst.insert({paramNames[i], val});
+    }
+    llvm::BasicBlock *block = llvm::BasicBlock::Create(Module->getContext(), "entry", applyRule);
+    llvm::BasicBlock *stuck = llvm::BasicBlock::Create(Module->getContext(), "stuck");
+    llvm::FunctionType *AbortType = llvm::FunctionType::get(llvm::Type::getVoidTy(Module->getContext()), false);
+    llvm::Function *AbortFunc = llvm::dyn_cast<llvm::Function>(Module->getOrInsertFunction("abort", AbortType));
+    AbortFunc->addFnAttr(llvm::Attribute::NoReturn);
```

Comment:
is this idempotent? This looks like it might pile on a lot of duplicate attributes.

Reply:
I mean the resulting llvm code doesn't have any duplicate noreturn attributes so I'm assuming yes.

Reply:
Looks like there's some deduplication in the implementation of attribute lists or AttrBuilder.